### PR TITLE
Perform I2C combined transactions when possible

### PR DESCRIPTION
**Pull request by cbeytas**
Perform I2C combined transactions whenever possible, within the
restrictions of the Broadcomm Serial Controller.

### CODE REVIEW

**Comment by popcornmix** [comment]
Can any I2C users test this patch and report?

**Comment by cbeytas** [comment]
Yes please. It should work for everyone. Limited testing done with an EEPROM and IR sensor.
Verified with a protocol analyzer that I2C message format is now correct.

https://i2c.wiki.kernel.org/index.php/Main_Page
http://www.raspberrypi.org/phpBB3/viewtopic.php?t=15840

**Comment by dsandber** [comment]
This patch worked for the MMA8451Q.  Can it please be included in the stock kernel?

**Comment by popcornmix** [comment]
@cbeytas 
Do you believe this patch will break any existing the driver for any existing users?

Any objections to including this?

**Comment by cbeytas** [comment]
It shouldn't break anything. I've tested it with a MLX90620 and its EEPROM, AT24C01A, PCF8591, and an adafruit LED backpack.

**Comment by popcornmix** [comment]
I've cherry-picked to 3.12.y kernel. It will be in next firmware update (later tonight).
Please test it still works okay on new kernel, and I'll keep an eye out for any reports of regressions.

**Comment by popcornmix** [comment]
@cbeytas, @dsandber and other I2C users. Please run rpi-update and test the behaviour is as desired.

**Comment by dsandber** [comment]
I'm confused.  I thought this was a kernel patch only -- I got my accelerometer working after I updated to the jjackowski patched kernel module.  How does this relate to the Pi firmware?

**Comment by popcornmix** [comment]
@dsandber rpi-update updates gpu firmware and linux kernel + modules. The updating kernel contains this patch.

**Comment by dsandber** [comment]
I confirm that after running rpi-update and rebooting my I2C sensor works properly.

**Comment by popcornmix** [comment]
This seems to have caused this regression:
https://github.com/raspberrypi/linux/issues/599

**Comment by dsandber** [comment]
Yikes.  Can we make this patch a module option until we figure out how to please everyone with the same code?

**Comment by popcornmix** [comment]
@dsandber A module option (defaulted to disabled) is one course of action.
I'd like to know if @cbeytas has any thoughts for a solution that doesn't require knowing about a module option.

**Comment by cbeytas** [comment]
I updated to the 3.12.20+ kernel (May 23) and my IR sensor no longer works. I'll have to troubleshoot why.

**Comment by popcornmix** [comment]
Due to problems reported, this is now disabled by default. Enable with module parameter combined=1.

**Comment by polppol** [comment]
This one also cause me a problem with PiCoolFan 
[For FW between 17-28may ,version between https://github.com/Hexxeh/rpi-firmware/commit/97082b6e95a2792660c2f79389748beccdbc568e and https://github.com/Hexxeh/rpi-firmware/commit/81355451bcd9a214fdf221ca322b6ca681d8da55]

and it's now working with
https://github.com/Hexxeh/rpi-firmware/commit/81355451bcd9a214fdf221ca322b6ca681d8da55

**Comment by dsandber** [comment]
My Pi is thoroughly locking-up when I read from my I2C MMA8451Q accelerometer continuously for a few minutes.  This used to work.  It happens both with and without the "combined" flag.  

I'd like to confirm this is a new problem by comparing to jjackowski's original module here, which worked:

http://jjackowski.wordpress.com/2014/01/08/replacement-i2c-kernel-module-for-raspbian/

However, when I downgrade the kernel using "rpi-update hash_of_older_version" and then do a modprobe on the original module I get an invalid module error.  Perhaps this is related to his module being for "3.10.25" while I downgraded to "3.10.25+"?  Anyone have any advice here?

In other words, I want to downgrade to a firmware/kernel that is compatible with his original module but I don't know how.  Thanks!

**Comment by dsandber** [comment]
I've experimented quite a bit reading my I2C accelerometer and found the following:

The newest (from rpi-update) firmware/kernel will LOCKUP eventually with combined=1
The newest will give bad data but not crash WITHOUT combined=1
The original jjackowski module works properly with the raspberrypi-bootloader package from 2014/1/7.

So it's still likely that we could make combined=1 the default once we fix the LOCKUP problem.

**Comment by dsandber** [comment]
Why does the cbeytas patch look totally different than the jjackowski one?  Can I merge the two together (so it's like the jjackowski one but his changes are disabled unless the "combined" flag is used)?  Because right now I'm pretty sure the combined flag will cause a lock-up for anyone if used long enough.
